src/Symfony/Component/String/AbstractUnicodeString.php

```diff
@@ -0,0 +1,450 @@
+                }, $s);
+            }
+        }
+
+        $str->string .= $s;
+
+        return $str;
+    }
+
+    public function codePoint(int $index = 0): ?int
+    {
+        $str = $offset ? $this->slice($index, 1) : $this;
+
+        return '' === $str->string ? null : mb_ord($str->string);
+    }
+
+    public function folded(bool $compat = true): parent
+    {
+        $str = clone $this;
+
+        if (!$compat || \PHP_VERSION_ID < 70300 || !\defined('Normalizer::FORM_KC_CF')) {
+            $str->string = normalizer_normalize($str->string, $compat ? \Normalizer::FORM_KC : \Normalizer::FORM_C);
+            $str->string = mb_strtolower(str_replace(self::FULL_CASE_FOLD_FROM, self::FULL_CASE_FOLD_TO, $this->string), 'UTF-8');
+        } else {
+            $str->string = normalizer_normalize($str->string, \Normalizer::FORM_KC_CF);
+        }
+
+        return $str;
+    }
+
+    public function join(array $strings): parent
+    {
+        $str = clone $this;
+        $str->string = implode($this->string, $strings);
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    public function lower(): parent
+    {
+        $str = clone $this;
+        $str->string = mb_strtolower(str_replace('İ', 'i̇', $str->string), 'UTF-8');
+
+        return $str;
+    }
+
+    public function match(string $pattern, int $flags = 0, int $offset = 0): array
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+            $match = 'preg_match_all';
+        } else {
+            $match = 'preg_match';
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (false === $match($pattern.'u', $this->string, $matches, $flags, $offset)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        return $matches;
+    }
+
+    /**
+     * @return static
+     */
+    public function normalize(int $form = self::NFC): self
+    {
+        if (!\in_array($form, [self::NFC, self::NFD, self::NFKC, self::NFKD])) {
+            throw new InvalidArgumentException('Unsupported normalization form.');
+        }
+
+        $str = clone $this;
+        normalizer_is_normalized($str->string, $form) ?: $str->string = normalizer_normalize($str->string, $form);
```

Comment:
So it is possible to get a GraphemeString instance with another form than `FORM_C`, is this correct?
Are the methods in GraphemeString still working correctly then? I'm asking because many methods are normalizing the input to `FORM_C`.

Reply:
> So it is possible to get a GraphemeString instance with a form other than FORM_C, is this correct?

Yes, but I've now made it normalize again back to NFC when being cloned!

> Are the methods in GraphemeString still working correctly then? I'm asking because many methods are normalizing the input to FORM_C.

This means string matching won't work that's correct, just for the normalized instance (since cloning will renormalize back to NFC now). That's expected since that's what one asked for (in this case, I'd expect normalization would be the last call in a chain of fluent transformations before being cast to another kind of string.)

Reply:
ok, fixed completely: lookups in NFD strings are now using NFD for needles too.
(note that an NFKD string *is* an NFD one too by definition)